File: 08. Trees/02. Binary Search Tree/02. Ceil and Floor in a BST/main.java
class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; left = right = null; }
}

public class main {
    // Iterative Ceil
    public static int findCeilIterative(TreeNode root, int key) {
        int ceil = -1;
        while (root != null) {
            if (root.val == key) return root.val;
            if (root.val > key) {
                ceil = root.val;
                root = root.left;
            } else {
                root = root.right;
            }
        }
        return ceil;
    }

    // Recursive Ceil
    public static int findCeilRecursive(TreeNode root, int key) {
        if (root == null) return -1;
        if (root.val == key) return root.val;
        if (root.val < key) return findCeilRecursive(root.right, key);
        int ceil = findCeilRecursive(root.left, key);
        return (ceil >= key) ? ceil : root.val;
    }

    // Iterative Floor
    public static int findFloorIterative(TreeNode root, int key) {
        int floor = -1;
        while (root != null) {
            if (root.val == key) return root.val;
            if (root.val < key) {
                floor = root.val;
                root = root.right;
            } else {
                root = root.left;
            }
        }
        return floor;
    }

    // Recursive Floor
    public static int findFloorRecursive(TreeNode root, int key) {
        if (root == null) return -1;
        if (root.val == key) return root.val;
        if (root.val > key) return findFloorRecursive(root.left, key);
        int floor = findFloorRecursive(root.right, key);
        return (floor <= key && floor != -1) ? floor : root.val;
    }

    // Main function
    public static void main(String[] args) {
        TreeNode root = new TreeNode(8);
        root.left = new TreeNode(4);
        root.right = new TreeNode(12);
        root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.right.left = new TreeNode(10);
        root.right.right = new TreeNode(14);

        int key = 5;
        System.out.println("Ceil (Iterative): " + findCeilIterative(root, key));
        System.out.println("Ceil (Recursive): " + findCeilRecursive(root, key));
        System.out.println("Floor (Iterative): " + findFloorIterative(root, key));
        System.out.println("Floor (Recursive): " + findFloorRecursive(root, key));
    }
}

File: 08. Trees/02. Binary Search Tree/02. Ceil and Floor in a BST/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Iterative Ceil
def find_ceil_iterative(root, key):
    ceil = -1
    while root:
        if root.val == key:
            return root.val
        if root.val > key:
            ceil = root.val
            root = root.left
        else:
            root = root.right
    return ceil

# Recursive Ceil
def find_ceil_recursive(root, key):
    if not root:
        return -1
    if root.val == key:
        return root.val
    if root.val < key:
        return find_ceil_recursive(root.right, key)
    
    ceil = find_ceil_recursive(root.left, key)
    return ceil if ceil >= key else root.val

# Iterative Floor
def find_floor_iterative(root, key):
    floor = -1
    while root:
        if root.val == key:
            return root.val
        if root.val < key:
            floor = root.val
            root = root.right
        else:
            root = root.left
    return floor

# Recursive Floor
def find_floor_recursive(root, key):
    if not root:
        return -1
    if root.val == key:
        return root.val
    if root.val > key:
        return find_floor_recursive(root.left, key)
    
    floor = find_floor_recursive(root.right, key)
    return floor if floor <= key and floor != -1 else root.val

# Main function
def main():
    root = TreeNode(8)
    root.left = TreeNode(4)
    root.right = TreeNode(12)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(6)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(14)

    key = 5
    print("Ceil (Iterative):", find_ceil_iterative(root, key))
    print("Ceil (Recursive):", find_ceil_recursive(root, key))
    print("Floor (Iterative):", find_floor_iterative(root, key))
    print("Floor (Recursive):", find_floor_recursive(root, key))

# Run the main function
if __name__ == "__main__":
    main()
